Программа повышения квалификации (научно-педагогических) работников НИУ ВШЭ

# Python для исследователей

# Задачи # 3

*Авторы: Татьяна Рогович*

## 1. Количество различных чисел (списки и множества)

Дан список чисел, который может содержать до 100000 чисел. Определите, сколько в нем встречается различных чисел.

#### Пример 1

##### Ввод	

1 2 3 2 1

##### Вывод

3

#### Пример 2
##### Ввод	

1 2 3 4 5 1 2 1 2 7 3

##### Вывод
6

In [1]:
# (∩｀-´)⊃━☆ﾟ.*･｡ﾟ
lst = input().split()
print(len(set(lst)))

6


## 2. Факультативы (множества)
Группа из 3 студентов пишет заявки на желаемые факультативы из списка: английский, немецкий, право, математика, сольфеджио. Факультатив откроют, если на него запишутся все студенты. Каждый студент может выбрать минимум один и максимум три факультатива. Нужно посчитать количество факультативов, которые откроются.

*Пример*

**Ввод:**

английский сольфеджио право  
математика сольфеджио  
немецкий право

**Вывод:**

0

**Ввод:**

математика немецкий право  
математика немецкий  
немецкий право математика 

**Вывод:**  
2

In [2]:
# считываем значения для каждого студента. input() принимает на ввод строку вида 'немецкий право математика'
# split() разбивает ее по пробелам и делает из нее список вида ['немецкий', 'право', 'математика'] (каждое слово, отделенное пробелом, стало объектом в нем)
# set() из этого списка делает множество (уникальные предметы, которые выбрал студент)
stud1 = set(input().split())
stud2 = set(input().split())
stud3 = set(input().split())
print(len(stud1 & stud2 & stud3)) # находим пересечение трех множеств (факультативы, которые общие для всех трех студентов) и смотрим, сколько объектов в этом множестве (len())


2


Давайте модифицируем эту задачу, и представим, что только два студента выбирают факультативы, откроются факультативы, которые выбрал, хотя бы один студент. Но затем выясняется, что один из факультативов из списка отменили. Выводить будет через запятую те факультативы, которые откроются.

*Пример*  
**Ввод:**

математика немецкий право  
математика английский  
право

**Вывод:**  
математика, немецкий, английский

In [3]:
# решение 1
stud1 = set(input().split())
stud2 = set(input().split())
cancelled = input()

opened = (stud1 | stud2)
opened.discard(cancelled)
print(*opened, sep=', ')

а, м, н


In [4]:
# решение 2
stud1 = set(input().split())
stud2 = set(input().split())
cancelled = set([input()])

opened = (stud1 | stud2) - cancelled
print(*opened, sep=', ')

а, м, н


## 3. Вложенные словари

Вам дан словарь, содержащий в себе списки словарей. Напишите код, который выведет на экран 'разработчик' из списка ключа Windows в в первом словаре из списка 2.

In [5]:
dict_of_lists = {'Список1': [{'Python': 'язык программирования'}, {'R':'язык программирования', 'LaTEX' : 'язык верстки' }], 
                 'Список2' : [{'Windows' : ['операционная система', 'разработчик'], 'UNIX' : 'операционная система'},
                              {'IBM': ['компания-производитель', 'разработчик'], 'IPv6' : 'интернет-протокол' }]}

# ваше решение здесь
dict_of_lists['Список2'][0]['Windows'][1]

'разработчик'

## 4. Лесенка

На вход принимается число N. 

Выведите на печать следующий паттерн с помощью цикла for:

1

1 2

1 2 3

1 2 3 4

...

1 ... N

In [6]:
# ваше решение здесь
n = int(input())

for i in range(n+1):
    print(*[x for x in range(1,i+1)])


1
1 2
1 2 3
1 2 3 4
1 2 3 4 5
1 2 3 4 5 6
1 2 3 4 5 6 7
1 2 3 4 5 6 7 8


## 5. Замена букв

Дана строка. Замените в ней все буквы А на # с помощью спискового включения.


In [7]:
phrase = "АБАЖУР, АБАЗИНСКИЙ, АБАЗИНЫ, АББАТ, АББАТИСА, АББАТСТВО"
# ваше решение здесь
''.join(['#' if i == 'А' else i for i in phrase])

'#Б#ЖУР, #Б#ЗИНСКИЙ, #Б#ЗИНЫ, #ББ#Т, #ББ#ТИС#, #ББ#ТСТВО'

## 6. Считаем письма
Давайте теперь поработаем с настоящим файлом и действительно посчитаем в нем слова. Мы загрузим метаданные почтового сервера университета Мичигана. И попробуем найти, с какого адреса ушло больше всего писем. Автор письма находится в строке, которая начинается с 'From '.

Бонус: выведите все адреса и количество отправленных писем, отсортированных по возрастанию количества писем

In [9]:
# импортируем библиотеку для доступа к файлам в интернете
import requests
# в переменной mbox хранится текст для работы
mbox = requests.get('http://www.py4inf.com/code/mbox.txt').text

In [11]:
# ваше решение здесь
lines = mbox.split('\n')
emails = {}

for line in lines:
    if line.startswith('From '):
        email = line.split()[1]
        if email in emails:
            emails[email] += 1
        else:
            emails[email] = 1
            
# выводим человека, который больше всего пишет
for email in emails:
    if emails[email] == max(emails.values()):
        print(email, emails[email])


zqian@umich.edu 195


In [12]:
# выводим все адреса по возрастанию количества отправленных писем
for value in sorted(set(emails.values())):
    for email in emails:
        if emails[email] == value:
            print(email, emails[email])

jlrenfro@ucdavis.edu 1
thoppaymallika@fhda.edu 1
colin.clark@utoronto.ca 1
jleasia@umich.edu 2
gbhatnag@umich.edu 3
bahollad@indiana.edu 4
knoop@umich.edu 5
ssmail@indiana.edu 5
tnguyen@iupui.edu 6
hu2@iupui.edu 7
lance@indiana.edu 8
john.ellis@rsmart.com 8
ggolden@umich.edu 8
jzaremba@unicon.net 9
mbreuker@loi.nl 9
ktsao@stanford.edu 12
a.fish@lancaster.ac.uk 14
kimsooil@bu.edu 14
stuart.freeman@et.gatech.edu 17
zach.thomas@txstate.edu 17
ostermmg@whitman.edu 17
antranig@caret.cam.ac.uk 18
csev@umich.edu 19
louis@media.berkeley.edu 24
gopal.ramasammycook@gmail.com 25
bkirschn@umich.edu 27
arwhyte@umich.edu 27
gsilver@umich.edu 28
josrodri@iupui.edu 28
nuno@ufp.pt 28
stephen.marquard@uct.ac.za 29
ray@media.berkeley.edu 32
sgithens@caret.cam.ac.uk 43
wagnermr@iupui.edu 44
gjthomas@iupui.edu 44
ajpoland@iupui.edu 48
david.horwitz@uct.ac.za 67
dlhaines@umich.edu 84
rjlowe@iupui.edu 90
jimeng@umich.edu 93
ian@caret.cam.ac.uk 96
aaronz@vt.edu 110
chmaurer@iupui.edu 111
cwen@iupui.edu 158
mm

### 7. Считаем метрику spam
Работаем с тем же файлом. У каждого письма есть метрика X-DSPAM-Confidence - уверенность фильтра в том, что это письмо не спам. Нужно посчитать среднее значение этой метрики для всей переписки.

In [13]:
# импортируем библиотеку для доступа к файлам в интернете
import requests
# в переменной mbox хранится текст для работы
mbox = requests.get('http://www.py4inf.com/code/mbox.txt').text
# ваше решение здесь

lines = mbox.split('\n')
spam = []

for line in lines:
    if line.startswith('X-DSPAM-Confidence'):
        spam.append(float(line.split()[1]))
        
print(f'Average X-DSPAM-Confidence: {round(sum(spam)/len(spam), 4)}')

Average X-DSPAM-Confidence: 0.8941


## 8. Объединение словарей

Напишите программу, которая объединяет значения из двух списков.

Ввод:

    shops = [{'товар': 'яблоки', 'количество': 400}, {'товар': 'конфеты', 'количество': 300}, {'товар': 'яблоки', 'количество': 750}]


Вывод:

    {'яблоки': 1150, 'конфеты': 300}

In [15]:
shops = [{'товар': 'яблоки', 'количество': 400}, 
         {'товар': 'конфеты', 'количество': 300}, 
         {'товар': 'яблоки', 'количество': 750}]

In [16]:
# ваше решение здесь
total = {}
for shop in shops:
    total[shop['товар']] = total.get(shop['товар'], 0) + shop['количество']
    
print(total)

{'яблоки': 1150, 'конфеты': 300}


## 9. Range

Найти все числа от 100 до 300 включительно, все цифры которых четные.

Программа должна выдавать результат в виде разделенной запятыми строки

In [18]:
# ваше решение здесь
print(*range(100, 301, 2), sep=',')

100,102,104,106,108,110,112,114,116,118,120,122,124,126,128,130,132,134,136,138,140,142,144,146,148,150,152,154,156,158,160,162,164,166,168,170,172,174,176,178,180,182,184,186,188,190,192,194,196,198,200,202,204,206,208,210,212,214,216,218,220,222,224,226,228,230,232,234,236,238,240,242,244,246,248,250,252,254,256,258,260,262,264,266,268,270,272,274,276,278,280,282,284,286,288,290,292,294,296,298,300


## 9. Считаем буквы и цифры 

Найти количество цифр и количество букв в строке

Вход: "абв532гд 2в34"


Выход:

    Цифры: 6
    Буквы: 6

In [21]:
letters = 0
digits = 0

for symbol in "абв532гд 2в34":
    if symbol.isdigit():
        digits += 1
    elif symbol.isalpha():
        letters += 1
        
print(letters)
print(digits)

6
6


# 10. Плэйлист

У Васи есть плейлист. Необходимо сгруппировать названия песен по исполнителям. 
Необходимо вывести исполнителей и песни, отсортированные по алфавиту 

**Формат ввода:**

Список исполнителей и песен в следующем формате:

* Число исполнителей
* Имя исполнителя: название песни 1, название песни 2 и т.д. 
* Имя каждого исполнителя и его песен вводится с новой строки 
* Названия песен перечислены через запятую 

**Формат вывода:**

Список исполнителей и песен, сгруппированных по исполнителю:

* Имя исполнителя: название песни 1, название песни 2 и т.д. 
* Имя каждого исполнителя и его песен вводится с новой строки 
* Названия песен перечислены через запятую 

*Пример:*

**Ввод:**

3

Queen: bohemian rhapsody, a kind of magic 

The beatles: and I love her 

Queen: the show must go on, bohemian rhapsody

**Вывод:**

Queen: a kind of magic, bohemian rhapsody, the show must go on

The beatles: and I love her

In [30]:
# ваше решение здесь
n = int(input())
music = {}

for line in range(n):
    item = input()
    singer, songs = item.split(': ')
    if singer in music:
        music[singer] |= set(songs.split(', '))
    else:
        music[singer] = set(songs.split(', '))
        
for key in music:
    print(f'{key}: {", ".join(music[key])}')

The beatles: and I love her 
Queen: a kind of magic , the show must go on, bohemian rhapsody


# 11. Оценки
 
**Пункт 1**  
Вам дан словарь с именами студентов и их оценками за курс по десятибаллной шкале. Напишите программу, которая бы считала среднюю оценку за курс и округляла ее (используйте функцию round()). Чтобы найти среднюю оценку нужно сложить оценки всех студентов за курс и разделить сумму на количество студентов. Найти сумму элементов списка, можно функцией sum(). Так, первый курс - это нулевой элемент списка оценок каждого студента.

Программа должна приниматься на ввод номер курса (от 1 до 8, обратите внимание, что не от 0 до 7), а выводить среднюю оценку за этот курс. 

**Пример ввода:**

1

**Пример вывода:**

Курс 1 - 4


In [31]:
marks = {'Mary' : [5, 8, 9, 10, 3, 5, 6, 6],
         'John' : [3, 3, 6, 8, 2, 1, 8, 5],
         'Alex' : [4, 4, 7, 4, 7, 3, 2, 9],
         'Patricia' : [2, 1, 6, 8, 2, 3, 7, 4]}

categories = {'отлично' : [8, 9, 10],
              'хорошо' : [6, 7],
              'удовлетворительно' : [4, 5],
              'неуд' : [0, 1, 2, 3]}

# Ваш код здесь
course = int(input('Введите номер курса от 1 до 8: ')) - 1

av = []
for student in marks:
    av.append(marks[student][course])
    
print(f'Курс {course + 1} - {round(sum(av)/len(av))}')

Курс 1 - 4


**Пункт 2**  
Напишите программу, которая бы сопостовляла оценку, найденную в предыдущем пункте, со вторым словарем, где оценкам в числах соответсвуют категории "хорошо", "отлично" и так далее. Программа принимает на ввод номер курса (от 1 до 8, обратите внимание, что не от 0 до 7), подсчитывает среднюю оценку за этот курс, а выводит буквенное значение оценки от "неуд" до "отлично" из словаря categories

**Пример ввода** 

1

**Пример вывода:**

Курс 1 - удовлетворительно


In [33]:
marks = {'Mary' : [5, 8, 9, 10, 3, 5, 6, 6],
         'John' : [3, 3, 6, 8, 2, 1, 8, 5],
         'Alex' : [4, 4, 7, 4, 7, 3, 2, 9],
         'Patricia' : [2, 1, 6, 8, 2, 3, 7, 4]}

categories = {'отлично' : [8, 9, 10],
              'хорошо' : [6, 7],
              'удовлетворительно' : [4, 5],
              'неуд' : [0, 1, 2, 3]}

# Ваш код здесь
course = int(input('Введите номер курса от 1 до 8: ')) - 1

av = []
for student in marks:
    av.append(marks[student][course])
    
mark_av = round(sum(av)/len(av))

for key in categories:
    if mark_av in categories[key]:
        print(f'Курс {course + 1} - {key}')

Курс 1 - удовлетворительно


**Пункт 3**  
Напишите программу, которая бы считала, сколько всего в словаре оценок, которые выше или равны определенному баллу.

Программа должна принимает на ввод оценку, а выводит количество оценок во всем словаре, которые больше этой оценки или равны ей. 

**Пример ввода:**

5

**Пример вывода:**

17


In [34]:
marks = {'Mary' : [5, 8, 9, 10, 3, 5, 6, 6],
         'John' : [3, 3, 6, 8, 2, 1, 8, 5],
         'Alex' : [4, 4, 7, 4, 7, 3, 2, 9],
         'Patricia' : [2, 1, 6, 8, 2, 3, 7, 4]}

categories = {'отлично' : [8, 9, 10],
              'хорошо' : [6, 7],
              'удовлетворительно' : [4, 5],
              'неуд' : [0, 1, 2, 3]}

# ваш код здесь
mark = int(input())
total = 0

for student in marks:
    for i in marks[student]:
        if i >= mark:
            total += 1
            
print(total)


17


# 12. Скидки в "Магнолии"

В магазине "Магнолия" проходит акция - даются скидки на товары, название которых начинается на заданную букву.

У Маши есть список покупок, состоящий из строк вида "цена, название, количество". Помогите ей посчитать, сколько денег она сэкономит (общую сумму скидки).

Количество записей в Машином списке покупок может быть любым. Каждая запись вводится с новой строки. Записи вводятся до тех пор, пока пользователь не введет "Конец списка". После этого одной строкой вводится буква, по которой сегодня определяется скидка. Если название продукта начинается на заданную букву, то на этот продукт действует скидка 15 процентов.

Образец вывода смотрите в примере. Если ни на один товар скидка не действует, то программа должна вывести "Скидка: 0 рублей"

*Вы считали категории в список списков, где каждая строка выглядит как в образце  - 1 балл*  
`[цена (int), название (str), количество (int)]`  
Например, `[120, мандарины (1 кг), 2]`

**Пример ввода**  
120 руб., мандарины (1 кг), 2 кг.  
70 руб., молоко, 3 пакета  
15 руб., батон сдобный, 1 штука  
345 руб., белое вино, 1 бутылка  
120 руб., моцарелла, 2 упаковки  
Конец списка  
м  

**Вывод**

Cкидка: 103.5 руб.


In [35]:
# ваше решение здесь
item = input()
shopping_list = []

while item != 'Конец списка':
    temp = []
    temp.append(int(item.split(', ')[0].split()[0]))
    temp.append(item.split(', ')[1])
    temp.append(item.split(', ')[2].split()[0])
    shopping_list.append(temp)
    item = input()
    
letter = input()
print(shopping_list, letter)

[[120, 'мандарины (1 кг)', '2'], [70, 'молоко', '3'], [15, 'батон сдобный', '1'], [345, 'белое вино', '1'], [120, 'моцарелла', '2']] м


In [36]:
discount = 0
for item in shopping_list:
    if item[1].startswith(letter):
        discount += item[0]*int(item[2])*0.15
        
print(f'Скидка: {discount} руб.')

Скидка: 103.5 руб.
